WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,243 @@
+	{
+		if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+		{
+			var url = urls.Where(url => url.Contains(".msi")).First();
+			return (url, url.Split("/").Last());
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+		{
+			var cpu = RuntimeInformation.ProcessArchitecture;
+			if (cpu.ToString() == "Arm64")
+			{
+				var arm64url = urls.Where(url => url.Contains("arm64.dmg")).First();
+				return (arm64url, arm64url.Split("/").Last());
+			}
+			var url = urls.Where(url => url.Contains(".dmg") && !url.Contains("arm64")).First();
+			return (url, url.Split("/").Last());
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+		{
+			throw new InvalidOperationException("For Linux, get the correct update manually.");
+		}
+		else
+		{
+			throw new InvalidOperationException("OS not recognized, download manually.");
+		}
+	}
+
+	public event EventHandler<UpdateStatus>? UpdateAvailableToGet;
+
+	public string DataDir { get; }
+	public string DownloadsDir { get; }
+	public IHttpClient HttpClient { get; }
+
+	///<summary>Comes from config file.</summary>
+	public bool DownloadNewVersion { get; set; }
+
+	///<summary>User's answer if installing new version on shutdown is needed.</summary>
+	public bool DoUpdateOnClose { get; set; } = false;
+
+	public bool InstallNewVersion()
+	{
+		try
+		{
+			var installerPath = Path.Combine(DataDir, "Downloads", InstallerName);
+			ProcessStartInfo startInfo;
+			if (!File.Exists(installerPath))
+			{
+				throw new FileNotFoundException(installerPath);
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				startInfo = ProcessStartInfoFactory.Make(installerPath, "", true);
+			}
+			else
+			{
+				startInfo = new()
+				{
+					FileName = installerPath,
+					UseShellExecute = true,
+					WindowStyle = ProcessWindowStyle.Normal
+				};
+			}
+
+			using Process? p = Process.Start(startInfo);
+
+			if (p is null)
+			{
+				throw new InvalidOperationException($"Can't start {nameof(p)} {startInfo.FileName}.");
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+			{
+				// For MacOS, you need to start the process twice, first start => permission denied
+				// TODO: find out why and fix.
+
+				p!.WaitForExit(5000);
+				p.Start();
+			}
+		}
+		catch (Exception ex)
+		{
+			Logger.LogError("Failed to install latest release. File might be corrupted.", ex);
+		}
+		return false;
+	}
+
+	public bool CheckIfInstallerDownloaded()
+	{
+		var folder = new DirectoryInfo(DownloadsDir);
+		if (folder.Exists)
+		{
+			FileSystemInfo[] files = folder.GetFileSystemInfos();
+
+			FileSystemInfo? file = files.Where(file => file.Name.Contains("Wasabi")).FirstOrDefault();
+
+			if (file is { } && file.Name.Contains("tmp"))
+			{
+				Logger.LogInfo("Corrupted/unfinished installer found, deleting.");
+				File.Delete(file.FullName);
+			}
+			else if (file is { })
+			{
+				InstallerName = file.Name;
+				return true;
+			}
```

Comment:
For me this is simpler.

```suggestion
			if (file is null)
			{
				Logger.LogInfo("Installer not found.");
				return false;
			}

			if (file.Name.Contains("tmp"))
			{
				Logger.LogInfo("Corrupted/unfinished installer found, deleting.");
				File.Delete(file.FullName);
			}
			else
			{
				InstallerName = file.Name;
				return true;
			}
```